d2/src/main/java/com/linkedin/d2/balancer/servers/ZooKeeperAnnouncer.java

```diff
@@ -211,8 +219,102 @@ public void onSuccess(None result)
         }
         runNextMarkUpOrMarkDown();
       }
-    });
+    };
+
+    final Callback<None> doMarkUpCallback = new Callback<None>() {
+      @Override
+      public void onError(Throwable e)
+      {
+        // this would ideally never happen
+        callback.onError(e);
+      }
+
+      @Override
+      public void onSuccess(None result)
+      {
+        _log.info("doing mark up of main cluster " + _cluster);
+        _server.markUp(_cluster, _uri, _partitionDataMap, _uriSpecificProperties, markUpCallback);
+      }
+    };
+
+    final Callback<None> warmupMarkDownCallback = new Callback<None>() {
+      @Override
+      public void onError(Throwable e) {
+        _log.info("mark down warm up cluster " + _warmupClusterName + " has failed");
+        // what to do here ? - go ahead continue to announce main cluster
+        // doAnnounceCallback.onSuccess(None.none());
+      }
+
+      @Override
+      public void onSuccess(None result) {
+        _log.info("mark down warm up cluster " + _warmupClusterName +
+            " has completed, now mark up of main cluster " + _cluster);
+        doMarkUpCallback.onSuccess(None.none());
+      }
+    };
+
+    final Callback<None> doWarmupMarkDownCallback = new Callback<None>() {
+      @Override
+      public void onError(Throwable e) {
+        _log.info("failure in scheduling warm up mark down");
+        // what to do here ? - go ahead continue to announce main cluster
+        // doAnnounceCallback.onSuccess(None.none());
+      }
+
+      @Override
+      public void onSuccess(None result) {
+        _log.info("warm up for cluster " + _warmupClusterName + " completed, now calling mark down on the warm up cluster");
+        _server.markDown(_warmupClusterName, _uri, warmupMarkDownCallback);
+      }
+    };
+
+    final Callback<None> doWarmupWaitCallback = new Callback<None>() {
+      @Override
+      public void onError(Throwable e)
+      {
+        _log.info("failure in mark up of warm up cluster " + _warmupClusterName);
+        // what to do here
+        // 1. go ahead continue to announce main cluster
+        // doAnnounceCallback.onSuccess(None.none());
+        // OR 2. mark announcement fail, move on to next pending mark up / mark down
+        // announceCallback.onError(e);
+      }
+
+      @Override
+      public void onSuccess(None result)
+      {
+        _log.info("mark up for warm up cluster " + _warmupClusterName + " succeeded");
+        ScheduledExecutorService warmup = Executors.newSingleThreadScheduledExecutor();
```

Comment:
In the last commit, I have added a new constructor which supplies ScheduledExecutorService.
Also added a setter method setExecutorServiceAndWarmupParameters to set _executorService and warm up parameters for the announcer. Using any of the 2 ways - the new constructor or the setter method, will need a change in container code, and the services that directly call methods on ZooKeeperAnnouncer if they have use to warmup. @rickzx @lesterhaynes please review and suggest.

Reply:
As per discussion with @rickzx  on slack, modified to only use the new constructor for ZooKeeperAnnouncer, because we don’t expect warm-up parameters to change at run-time.